firedrake/dmhooks.py

```diff
@@ -429,6 +429,7 @@ def coarsen(dm, comm):
                  teardown=partial(pop_appctx, cdm, cctx),
                  call_setup=True)
         # Necessary for MG inside a fieldsplit in a SNES.
+        dm.setKSPComputeOperators(firedrake.solving_utils._SNESContext.compute_operators)
```

Comment:
This is a bit mysterious! Can you add a comment explaining why?

Reply:
I'll add it in post. Basically there are hooks that are performed by petsc that attempt to carry compute operators around from outer to inner dms. The outer DM has a kspcomputeoperators_snes attached to it (which is wrong inside a fieldsplit for complex reasons) and that gets transferred onto the coarse DM after we've set up our compute operators. This way we make sure the fine DM has the right compute operators callback.